pkg/agent/controller/egress/egress_controller.go

```diff
@@ -466,6 +538,32 @@ func (c *EgressController) syncEgress(egressName string) error {
 		return err
 	}
 
```

Comment:
I think deleting Egress is not handled properly. The EgressIP is left on Node after I remove the egress. There should be some code in the above section.

Reply:
Does the agent need to remove the Egress IP?  How about the manually assigned IP?
1.Where an IP has been assigned by agent, the agent must delete it from the local node.
2.When an IP has been assigned manually, the agent does not need to delete it.

Anyway, I think we need to add a parameter check(EgressIP) here when deleting an Egress.

Reply:
It definitely needs to remove the Egress IP if it assigns it previously. Otherwise how the datapath works when an IP is on multiple Nodes?

It should delete the IP only if the IP is configured by agent. Do not touch user configured IPs.

I think instead of having too much logic in egress controller, you could just expose two methods of IPAssigner to controller:
`AssignEgressIP(egressName, egressIP)` and `UnassignIP(egressName)`.
AssignEgressIP will be called when the Egress has an IP should be assigned. It should take care of the following:
1. If the desired IP is already configured, do nothing.
2. If the egress had an IP which is different from the desired one, remove the previous one and configure the new one.
3. If the egress didn't have an IP, configure the new one.

UnassignEgressIP will be called when the Egress is removed or should no longer hold the IP. It should take care of the following:
1. If the egress had an IP, remove it.
2. If the egress didn't have an IP, do nothing.

In this way, the methods are idempotent and will make controller much simpler, which is the point of encapsulation.

Reply:
```
// IPAssigner provides methods to assign or unassign egressIP.
type IPAssigner interface {
	AssignEgressIP(egressIP, egressName string) error
	UnassignEgressIP(egressName string) error
	AssignedIPs() (ips map[string]string)
}
```

We need AssignedIPs to list all assigned EgressIP.